content/md/en/docs/reference/frame-macros.md

```diff
@@ -1,118 +1,206 @@
+  - `impl_opaque_keys` used to implement `OpaqueKeys` for a described data structure.
+  - `parameter_types` used to create new implementations of the `Get` trait.
+
+  For more information about `sp_runtime` function-like macros, see [Macros](https://paritytech.github.io/substrate/master/sp_runtime/index.html#macros).
+  For information about `sp_runtime` derive macros, see [Derive macros](https://paritytech.github.io/substrate/master/sp_runtime/index.html#derives).
+
+- [`sp_api`](https://paritytech.github.io/substrate/master/sp_api/index.html)
+  
+  - `decl_runtime_apis`	used to declare specified traits as runtime APIs.
+  - `impl_runtime_apis`	used to tag specified trait implementations as runtime APIs.
+  
+  For more information about `sp_api` function-like macros, see [Macros](https://paritytech.github.io/substrate/master/sp_api/index.html#macros).
+
+- [`sp_std`](https://paritytech.github.io/substrate/master/sp_std/index.html)
+  
+  - `if_std`	used to indicate code that should only be run when the `std` feature set is enabled.
+  - `map`	used to initialize a key-value collection from array.
+  - `vec`	used to create a vector containing the arguments.
+  
+  For more information about `sp_std` function-like macros, see [Macros](https://paritytech.github.io/substrate/master/sp_std/index.html#macros).
+
+- [`sp_version`](https://paritytech.github.io/substrate/master/sp_version/index.html)
+  
+  - `create_apis_vec`	used to create a vector of API declarations.
+  - `create_runtime_str` used to create a `RuntimeString` constant.
+  - `runtime_version`	used as an attribute that accepts the version declaration of a runtime and generates a custom WebAssembly section with the equivalent contents.
+
+You'll see these many of these crates listed as dependencies in the runtime and node `Cargo.toml` file for the node template.
+
+## Macros for composing pallets
 
 As discussed in [Building custom pallets](/main-docs/fundamentals/runtime-intro#building-custom-pallets), most FRAME pallets are composed using a common set of sections.
 Macros make building each of those sections more modular and extensible.
 This section describes the macros available and how to use them to build your custom runtime.
 
-### #[frame_support::pallet]
+### #[pallet]
 
-**When to use**
+The `#[pallet]` macro is required to declare a pallet.
+This [attribute macro](https://paritytech.github.io/substrate/master/frame_support/attr.pallet.html) describes the attributes used to identify the specific items the pallet requires.
+For example, a pallet typically includes a set of types, functions, and trait implementations that are aggregated by the `construct_runtime!` macro to build the runtime.
 
-Required to declare a pallet consisting of a set of types, functions, and trait implementations that will later be aggregated by `construct_runtime` to build to runtime.
+The `#[pallet]` attribute macro defines a pallet as a module:
```

Comment:
this sentence does not make sense to me

Reply:
macros can only act on code within some structure. the `#[pallet]` macro is an attribute on a `mod pallet`, which is an entry point for all the other `#[pallet::xx]` macros to act on other parts of code.

Reply:
The fact we wrap the code in a `mod pallet` is just an artifact of how we can make attribute macros to work

Reply:
Ah, that's much clearer than what I was trying to interpret from the Rust doc.